### PR TITLE
[llvm-exegesis] Align loop MBB in loop repetitor

### LINKED ISSUE #77259: [llvm-exegesis] Measurement values sometimes doubled
Currently, there is a transient failure in `llvm-exegesis` where the measurement values that are reported are occasionally double what they should be. For example, take the following invocation:
```shell
llvm-exegesis -mode=latency -opcode-name=ADD64rr -execution-mode=subprocess -repetition-mode=loop
```
Typically something like the following will be reported:
```yaml
measurements:
  - { key: latency, value: 1.006, per_snippet_value: 1.006 }
```
However, occasionally (on maybe 1 out of every 10 or 20 runs), the following is reported:
```yaml
measurements:
  - { key: latency, value: 2.0051, per_snippet_value: 2.0051 }
```
All the performance counter reads are pretty similar to each other, just double of what they should be:
```
20103
20107
20104
20104
20103
20113
20104
20783
20103
20103
20103
20107
20097
20104
20103
20103
20103
20103
20103
20104
20065
20104
20103
20103
20104
20104
20104
20107
20116
20051
```
(Obtained by manually adding a print statement in `PerfHelper.cpp`)

Given that all these measurements are double what they should be, it seems to point to a systematic failure within `llvm-exegesis`.

There don't seem to be any issues with the generated code:
```x86
      ee: 49 b8 10 27 00 00 00 00 00 00 movabsq $0x2710, %r8            # imm = 0x2710
      f8: 4d 01 e4                      addq    %r12, %r12
      fb: 49 83 c0 ff                   addq    $-0x1, %r8
      ff: 75 f7                         jne     0xf8 <foo+0xf8>
```
(`0x2710` is the same is 10000, as expected for the default `--num-repetitions` value of 10000).

Interestingly, this only seems to occur in the subprocess execution mode. I have not been able to reproduce this with in the inprocess execution mode (at least not yet).

Additionally, this seems like it might be more systematic on certain platforms. On Cascade Lake, I am consistently getting measurements higher than they should be rather than sporadically. Not sure if that is related however.

### PR DESCRIPTION
**Pull request by boomanaiden154**
This patch sets the alignment of the loob MBB in the loop repetitor to 16 to avoid instruction fetch/predecoding bottlenecks that can come up with unaligned code. The value of 16 was chosen based on numbers for recent Intel microarchitectures and reccomendations from Agner Fog.

Fixes #77259.

### CODE REVIEW

File: llvm/tools/llvm-exegesis/lib/SnippetRepetitor.cpp [hunk 1]
```diff
@@ -74,6 +74,10 @@ class LoopSnippetRepetitor : public SnippetRepetitor {
       auto Loop = Filler.addBasicBlock();
       auto Exit = Filler.addBasicBlock();
 
+      // Align the loop machine basic block to a sixteen byte boundary
+      // so that instruction fetch on modern x86 platforms works optimally.
+      Loop.MBB->setAlignment(Align(16));
```
Comment (legrosbuffle) [inline llvm/tools/llvm-exegesis/lib/SnippetRepetitor.cpp:79]: This should probably be using `TargetLowering::getPrefLoopAlignment`

**Review by legrosbuffle — APPROVED** [review]

File: llvm/tools/llvm-exegesis/lib/SnippetRepetitor.cpp (hunk 1 above)
Comment (boomanaiden154) [reply llvm/tools/llvm-exegesis/lib/SnippetRepetitor.cpp]: Thanks for the suggestion! That's definitely a lot cleaner. Currently X86 segfaults on this without passing a `MachineLoop` pointer which would be pretty difficult to get here. I've opened #77283 to fix the API issue. I'll land this patch after that one lands.

**Review by RKSimon — APPROVED** [review]
LGTM - cheers

**Comment by phoebewang** [comment]
Not quite understand the background, but want to share to see if the recent feature #71026 helps with this.

**Comment by boomanaiden154** [comment]
> Not quite understand the background, but want to share to see if the recent feature #71026 helps with this.

Thank you for pointing this out. We don't lower anything from LLVM IR and generate machine code directly for benchmarks in `llvm-exegesis`, so this isn't applicable with the way we're doing things currently.

**Pull request merged**